### PR TITLE
[feature](Load)Suppot MySQL Load Data

### PR DESCRIPTION
**Pull request by SaintBacchus**
# Proposed changes

Main subtask of [DSIP-28](https://cwiki.apache.org/confluence/display/DORIS/DSIP-028%3A+Suppot+MySQL+Load+Data)

## Problem summary
Support mysql load syntax as below: 
```sql
LOAD DATA
    [LOCAL]
    INFILE 'file_name'
    INTO TABLE tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [COLUMNS TERMINATED BY 'string']
    [LINES TERMINATED BY 'string']
    [IGNORE number {LINES | ROWS}]
    [(col_name_or_user_var [, col_name_or_user_var] ...)]
    [SET (col_name={expr | DEFAULT} [, col_name={expr | DEFAULT}] ...)]
    [PROPERTIES (key1 = value1 [, key2=value2]) ]
```

For example, 
```sql
            LOAD DATA 
            LOCAL
            INFILE 'local_test.file'
            INTO TABLE db1.table1
            PARTITION (partition_a, partition_b, partition_c, partition_d)
            COLUMNS TERMINATED BY '\t'
            (k1, k2, v2, v10, v11)
            set (c1=k1,c2=k2,c3=v10,c4=v11)
            PROPERTIES ("auth" = "root:", "strict_mode"="true")
```

Note that in this pr the property named `auth` must be set since stream load need auth. I will optime it later.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.33 seconds
 load time: 633 seconds
 storage size: 17123622841 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221230084431_clickbench_pr_71571.html

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadManager.java [hunk 1]
```diff
@@ -151,6 +181,183 @@ public void createLoadJobV1FromMultiStart(String fullDbName, String label) throw
         }
     }
 
+    public LoadJobRowResult executeMySqlLoadJobFromStmt(ConnectContext context, LoadStmt stmt)
+            throws IOException, LoadException {
+        LoadJobRowResult loadResult = new LoadJobRowResult();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadManager.java:186]: How about moving these methods to a separate class to make LoadManager simple?

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 2]
```diff
@@ -1836,6 +1841,41 @@ private void handleExplainStmt(String result) throws IOException {
...
+                throw new DdlException("Unknown load job type");
+            }
+            if (jobType == EtlJobType.HADOOP) {
+                throw new DdlException("Load job by hadoop cluster is disabled."
+                        + " Try using broker load. See 'help broker load;'");
+            }
+            LoadManager loadManager = context.getEnv().getLoadManager();
+            if (jobType == EtlJobType.LOCAL_FILE) {
+                if (!context.getCapability().isClientLocalFile()) {
+                    throw new DdlException("Doris server does not support load local file from mysql client.");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:1858]: The error msg is confusing.
The `if` says `not a client local file`, but err msg says "not support local file from client"?

File: fe/fe-core/src/main/jflex/sql_scanner.flex [hunk 3]
```diff
@@ -479,9 +479,13 @@ import org.apache.doris.qe.SqlModeHelper;
         keywordMap.put("year", new Integer(SqlParserSymbols.KW_YEAR));
         keywordMap.put("mtmv", new Integer(SqlParserSymbols.KW_MTMV));
         keywordMap.put("histogram", new Integer(SqlParserSymbols.KW_HISTOGRAM));
+<<<<<<< HEAD
         keywordMap.put("auto", new Integer(SqlParserSymbols.KW_AUTO));
         keywordMap.put("prepare", new Integer(SqlParserSymbols.KW_PREPARE));
         keywordMap.put("execute", new Integer(SqlParserSymbols.KW_EXECUTE));
+=======
```
Comment (morningman) [inline fe/fe-core/src/main/jflex/sql_scanner.flex:486]: Conflict

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java (hunk 2 above)
Comment (SaintBacchus) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java]: use a `1148` error code of these case

File: fe/fe-common/src/main/java/org/apache/doris/common/io/ByteBufferNetworkInputStream.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public ByteBufferNetworkInputStream(int capacity) {
+        this.queue = new ArrayBlockingQueue<>(capacity);
+    }
+
+    public void fillByteBuffer(ByteBuffer buffer) throws IOException, InterruptedException {
+        if (closed) {
+            throw new IOException("Stream is already closed.");
+        }
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(buffer.array(), buffer.position(), buffer.limit());
+        queue.offer(inputStream, 300, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (SaintBacchus) [reply fe/fe-common/src/main/java/org/apache/doris/common/io/ByteBufferNetworkInputStream.java:46]: It is an big time but not too big time for timeout balancing the  load success rate and the cluster overload.

[the remark this answers is not in the document]
Comment (SaintBacchus) [reply fe/fe-common/src/main/java/org/apache/doris/common/io/ByteBufferNetworkInputStream.java:46]: It's by experience

**Review by morningman — APPROVED** [review]
LGTM

**Comment by jackwener** [comment]
Why use `fastjson`? it's just used for dependence of `com.alibaba.otter:canal.protocol:jar:1.1.6:compile`

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+import org.apache.doris.common.io.ByteBufferNetworkInputStream;
+import org.apache.doris.load.LoadJobRowResult;
+import org.apache.doris.load.loadv2.LoadTask.MergeType;
+import org.apache.doris.qe.ConnectContext;
+import org.apache.doris.system.Backend;
+import org.apache.doris.system.BeSelectionPolicy;
+import org.apache.doris.system.SystemInfoService;
+
+import com.alibaba.fastjson2.JSON;
+import com.alibaba.fastjson2.JSONObject;
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:35]: Please replace them, avoid use fastjson in the code of Doris

Comment (SaintBacchus) [reply fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:35]: OK, I will remove it later.

**Comment by SaintBacchus** [comment]
@jackwener I had open a PR to disable using fastjson in fe #16235, can you help to review it?

**Pull request merged**